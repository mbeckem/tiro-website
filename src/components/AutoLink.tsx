import Link from "next/link";
import React from "react";
import { ExternalLink } from "./ExternalLink";

interface AutoLinkProps {
    href: string;
    className?: string;
}

/**
 * A simple link component that automatically uses next links for internal urls and ExternalLink otherwise.
 */
export const AutoLink: React.FC<AutoLinkProps> = (props) => {
    const href = props.href;
    if (/^#/.test(href)) {
        return <a {...props} />;
    }
    if (/^($|\/|\.\.?\/)/.test(href)) {
        const otherProps: any = Object.assign({}, props);
        delete otherProps.href;

        return (
            <Link href={href}>
                <a {...otherProps} />
            </Link>
        );
    }

    return <ExternalLink {...props} />;
};
