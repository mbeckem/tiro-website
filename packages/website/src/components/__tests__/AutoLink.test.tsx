import React from "react";
import { render, screen } from "@testing-library/react";
import { AutoLink } from "../AutoLink";

// Example test
describe("AutoLink", () => {
    it("renders an external link for full URLs", () => {
        render(<AutoLink className="test-class" href="https://example.com/foo">Hello World</AutoLink>);

        const link = screen.getByRole("link", {
            name: /Hello World/i
        });

        expect(link).toMatchSnapshot();
    });

    it("renders a normal for relative hrefs", () => {
        render(<AutoLink className="test-class" href="./foo/bar">Hello World</AutoLink>);

        const link = screen.getByRole("link", {
            name: /Hello World/i
        });

        expect(link).toMatchSnapshot();
    });

    it("renders a normal for anchor hrefs", () => {
        render(<AutoLink className="test-class" href="#anchor-123">Hello World</AutoLink>);

        const link = screen.getByRole("link", {
            name: /Hello World/i
        });

        expect(link).toMatchSnapshot();
    });
});
