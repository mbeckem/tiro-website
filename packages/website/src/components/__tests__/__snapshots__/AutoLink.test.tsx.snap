// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AutoLink renders a normal for anchor hrefs 1`] = `
<a
  class="test-class"
  href="#anchor-123"
>
  Hello World
</a>
`;

exports[`AutoLink renders a normal for relative hrefs 1`] = `
<a
  class="test-class"
  href="/foo/bar"
>
  Hello World
</a>
`;

exports[`AutoLink renders an external link for full URLs 1`] = `
<a
  class="test-class"
  href="https://example.com/foo"
  rel="noopener noreferrer"
  target="_blank"
>
  Hello World
  <span
    class="bp3-icon bp3-icon-share"
    icon="share"
    style="vertical-align: baseline; padding-left: 4px;"
  >
    <svg
      data-icon="share"
      height="12"
      viewBox="0 0 16 16"
      width="12"
    >
      <desc>
        share
      </desc>
      <path
        d="M10.99 13.99h-9v-9h4.76l2-2H.99c-.55 0-1 .45-1 1v11c0 .55.45 1 1 1h11c.55 0 1-.45 1-1V7.24l-2 2v4.75zm4-14h-5c-.55 0-1 .45-1 1s.45 1 1 1h2.59L7.29 7.28a1 1 0 00-.3.71 1.003 1.003 0 001.71.71l5.29-5.29V6c0 .55.45 1 1 1s1-.45 1-1V1c0-.56-.45-1.01-1-1.01z"
        fill-rule="evenodd"
      />
    </svg>
  </span>
</a>
`;
